### PR TITLE
misc: add flexibility to delete_ and move_file

### PR DESCRIPTION
**Pull request by jan--f**
Add check argument to delete_file to provide the ability to ignore a
failed delete (say if the file to delete does not exist)
Add preserve_perms argument to move_file to make the preservation of
permissions optional. This makes move_file usable if the target
file does not yet exist.
Existing invocations are unchanged.

Signed-off-by: Jan Fajerski <jfajerski@suse.com>
(cherry picked from commit f9aef0126313a24b791bc16918eb47d113ee6139)

### CODE REVIEW

**Comment by dmick** [comment]
a couple of minor issues so far: 1) try not to introduce new pep8 issues (I see a couple of instances of if(condition))  2) in the commit message, you typo "preserver_parms" instead of "preserve_parms"

**Comment by dmick** [comment]
ok to test

**Comment by jan--f** [comment]
@dmick thx...fixed.

**Comment by smithfarm** [comment]
@jan--f: Can you drop the "cherry-picked from" line, too?

**Comment by jan--f** [comment]
@smithfarm dropped.

File: teuthology/misc.py [hunk 1]
```diff
@@ -586,22 +590,23 @@ def move_file(remote, from_path, to_path, sudo=False):
...
+        if sudo:
+            args.append('sudo')
+        args.extend([
+            'chmod',
+            perms,
+            to_path,
+        ])
+        proc = remote.run(
+            args=args,
+            stdout=StringIO(),
```
Comment (dmick) [inline teuthology/misc.py]: but you're not consuming stdout, so why set it up to a StringIO() object?

File: teuthology/misc.py [hunk 2]
```diff
@@ -586,22 +590,23 @@ def move_file(remote, from_path, to_path, sudo=False):
...
+        )
 
 
-def delete_file(remote, path, sudo=False, force=False):
+def delete_file(remote, path, sudo=False, force=False, check=True):
```
Comment (dmick) [inline teuthology/misc.py]: Why do we need this given that we have force, which ignores "file not found" errors as well?

File: teuthology/misc.py [hunk 3]
```diff
@@ -550,27 +550,31 @@ def copy_file(from_remote, from_path, to_remote, to_path=None):
...
+    clobbered if it does.
+    """
+    if preserve_perms:
+        args = []
+        if sudo:
+            args.append('sudo')
+        args.extend([
+            'stat',
+            '-c',
+            '\"%a\"',
```
Comment (dmick) [inline teuthology/misc.py]: Are the backslashes for the further shell escape?  This is what "run.Raw()" is for, but I also don't think the quotes are necessary:

>>> r.run(args=['echo', '%a'])
2017-03-20 09:13:46,118.118 INFO:teuthology.orchestra.run.mira001:Running: 'echo %a'
2017-03-20 09:13:46,302.302 INFO:teuthology.orchestra.run.mira001.stdout:%a

Edit: I see you were just moving code.  Still.

File: teuthology/misc.py [hunk 4]
```diff
@@ -550,27 +550,31 @@ def copy_file(from_remote, from_path, to_remote, to_path=None):
...
+            to_path
+        ])
+        proc = remote.run(
+            args=args,
+            stdout=StringIO(),
+        )
+        perms = proc.stdout.getvalue().rstrip().strip('\"')
```
Comment (dmick) [inline teuthology/misc.py]: ...which means you wouldn't need this stripping here too:

```
>>> r.run(args=['stat', '-c', '%a', '/tmp'])
2017-03-20 09:15:16,281.281 INFO:teuthology.orchestra.run.mira001:Running: 'stat -c %a /tmp'
2017-03-20 09:15:16,466.466 INFO:teuthology.orchestra.run.mira001.stdout:1777

```

File: teuthology/misc.py [hunk 5]
```diff
@@ -550,27 +550,31 @@ def copy_file(from_remote, from_path, to_remote, to_path=None):
     ])
 
 
-def move_file(remote, from_path, to_path, sudo=False):
+def move_file(remote, from_path, to_path, sudo=False, preserve_perms=True):
```
Comment (dmick) [inline teuthology/misc.py]: So the description says the desire is to make this workable even if the destination doesn't exist, but, the method could simply ignore the 'no perms to update' without requiring the caller to request it.  The issue is, however, what permissions for the destination file ought to be used (whatever the source has, or some default, or?...)

I think I'd be happier if the change here were to add a permission for the case of the file not existing (as an optional kw parameter, still).  If to_path doesn't exist, then use perm if supplied, otherwise default to the perm of from_path.  What do you think?

File: teuthology/misc.py (hunk 2 above)
Comment (jan--f) [inline teuthology/misc.py]: True, I did not consider using -f. Will update.

File: teuthology/misc.py (hunk 3 above)
Comment (jan--f) [inline teuthology/misc.py]: Yeah I just indented existing code. I see your points, but shouldn't this be addressed in a separate PR?

File: teuthology/misc.py (hunk 5 above)
Comment (jan--f) [inline teuthology/misc.py]: I think it is better this way tbh. ```move_file``` already does quite a bit more then its name suggests. Starting to now also set permissions (but only if ```to_path``` doesn't exist) seems to add complexity beyond this functions scope.
Personally I'd expect it to behave like ```mv```, so preserve permissions of the source file. I added the option to keep things compatible with existing code.

File: teuthology/misc.py (hunk 3 above)
Comment (dmick) [inline teuthology/misc.py]: yep.

File: teuthology/misc.py (hunk 5 above)
Comment (dmick) [inline teuthology/misc.py]: I guess that's the part that rankles a bit; it's changing the intent of the original definition, however misguided, while leaving the default behavior, but not going the step further to add the capability of choosing the actual permission.  But I can see that it's certainly a matter of opinion which way is more in keeping with expectations.  I withdraw the suggestion.

**Comment by dmick** [comment]
Sorry I let this drop @jan--f.  @zmc, if you're ok with it I'm ok.
